master/buildbot/changes/github.py

```diff
@@ -0,0 +1,259 @@
+
+    @defer.inlineCallbacks
+    def _getCurrentRev(self, prnumber):
+        # Get currently assigned revision of PR number
+
+        result = yield self._getStateObjectId()
+        rev = yield self.master.db.state.getState(result, 'pull_request%d' %
+                                                  prnumber, None)
+        defer.returnValue(rev)
+
+    @defer.inlineCallbacks
+    def _setCurrentRev(self, prnumber, rev):
+        # Set the updated revision for PR number.
+
+        result = yield self._getStateObjectId()
+        yield self.master.db.state.setState(result,
+                                            'pull_request%d' % prnumber, rev)
+
+    @defer.inlineCallbacks
+    def _getStateObjectId(self):
+        # Return a deferred for object id in state db.
+        result = yield self.master.db.state.getObjectId(
+            '%s/%s' % (self.owner, self.repo), self.db_class_name)
+        defer.returnValue(result)
+
+    @defer.inlineCallbacks
+    def _processChanges(self, github_result):
+        for pr in github_result:
+            # Track PRs for specified branches
+            base_branch = pr['base']['ref']
+            prnumber = pr['number']
+            revision = pr['head']['sha']
+
+            # Check to see if the branch is set or matches
+            if base_branch not in self.branches:
+                return
+            if self.pullrequest_filter and not self.pullrequest_filter(pr):
+                return
+            current = yield self._getCurrentRev(prnumber)
+            if not current or current[0:12] != revision[0:12]:
+                # Access title, repo, html link, and comments
+                branch = pr['head']['ref']
+                title = pr['title']
+                repo = pr['head']['repo']['name']
+                revlink = pr['html_url']
+                comments = pr['body']
+                if self.useTimestamps:
+                    updated = datetime.strptime(pr['updated_at'],
+                                                '%Y-%m-%dT%H:%M:%SZ')
+                else:
+                    updated = epoch2datetime(reactor.seconds())
+
+                # update database
+                yield self._setCurrentRev(prnumber, revision)
+
+                author = pr['user']['login']
+
+                dl = defer.DeferredList(
+                    [self._getFiles(prnumber), self._getEmail(author)],
+                    consumeErrors=True)
+
+                results = yield dl
+                failures = [r[1] for r in results if not r[0]]
+                if failures:
+                    for failure in failures:
+                        log.err(failure, "while processing changes for "
+                                "Pullrequest {} revision {}".format(prnumber,
+                                                                    revision))
+                        # Fail on the first error!
+                        failures[0].raiseException()
```

Comment:
probably we need a test on that too.
What are we doing in that case? are we retrying the same PR in the next poll?